Allow passing cpu to CUDA RPC device maps

beauby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57019 Allow passing cpu to CUDA RPC device maps**

Based on https://github.com/pytorch/pytorch/pull/56043

Differential Revision: [D28169796](https://our.internmc.facebook.com/intern/diff/D28169796)

beauby committed:
Allow passing cpu to CUDA RPC device maps

Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

lw commented on torch/distributed/rpc/options.py:
```diff
@@ -25,19 +25,16 @@ def _to_device_index_map(device_map: Dict[DeviceType, DeviceType]) -> Dict[int,
+        if v_index in reverse_map:
             raise ValueError(
                 "`device_map` only supports 1-to-1 mapping, "
                 f"trying to map {k} and {reverse_map[v.index]} to {v.index}"
```
```suggestion
                f"trying to map {k} and {reverse_map[v_index]} to {v_index}"
```
I'll soon send a PR which will make us use "complete" devices here, rather than just indices, and then we'll be able to easily print "cpu" instead of "-1", but for now this might be fine.

lw reviewed (approved):
Apart from those timeouts, this looks good to me!

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -489,12 +489,6 @@ def remote(to, func, args=None, kwargs=None, timeout=UNSET_RPC_TIMEOUT):
         value. Use the blocking API :meth:`torch.distributed.rpc.RRef.to_here`
         to retrieve the result value locally.
 
-    .. warning ::
-        Using GPU tensors as arguments or return values of ``func`` is not
-        supported since we don't support sending GPU tensors over the wire. You
-        need to explicitly copy GPU tensors to CPU before using them as
-        arguments or return values of ``func``.
```
Yep, this is indeed obsolete. It was added in https://github.com/pytorch/pytorch/commit/f058c03b1584e3dc50d0e2abd79c7ebb646c8125. Thanks for removing it!

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM too!

beauby committed:
Update on "Allow passing cpu to CUDA RPC device maps"


Based on https://github.com/pytorch/pytorch/pull/56043

Differential Revision: [D28169796](https://our.internmc.facebook.com/intern/diff/D28169796)

[ghstack-poisoned]
